### ISSUE TITLE
[BUG]  unit test  AtomStyles Failed (AtomStyleTest.body_nparticle)

### ISSUE THREAD

**Issue by iridos** [lammps/lammps#2636]
**Summary**
I compiled the serial version of stable_29Oct2020 from git using the intel compiler 19.1  and mkl 2020 on RHEL7 and run the unit tests with "ctest".
The AtomStyles test failed. Verbose/debug output shows that the results calculated differ from expectations for AtomStyleTest.body_nparticle

OS: RHEL7
Hardware: Intel(R) Xeon(R) Gold 6252 CPU @ 2.10GH
compiler: intel 19.1
numerical lib: mkl 2020

$ ./lmp_serial 
LAMMPS (29 Oct 2020)

Full output on the failed test:
$ ctest -VV  -d  --rerun-failed                                                                           
  (edit: removed because formatting as "code" removed newlines. Adding as attachment)
[ctest-failed.log](https://github.com/lammps/lammps/files/6061791/ctest-failed.log)

Packages compiled into LAMMPS:
Enabled packages: ASPHERE;BODY;CLASS2;COLLOID;COMPRESS;CORESHELL;DIPOLE;GRANULAR;KIM;KOKKOS;KSPACE;MANYBODY;MC;MISC;MOLECULE;OPT;PERI;POEMS;PYTHON;QEQ;REPLICA;RIGID;SHOCK;SNAP;SRD;USER-COLVARS;USER-DRUDE;USER-EFF;USER-INTEL;USER-OMP;USER-REAXC

Compilier warnings/remarks
(listed for completeness. only one shown of each type)
* /usr/include/python2.7/pyconfig-64.h(1232): warning #47: incompatible redefinition of macro "_POSIX_C_SOURCE" (declared at line 265 of "/usr/include/features.h"
* …USER-INTEL/pair_airebo_intel.cpp(538): (col. 7) remark: Loop with pragma of unroll = 4 ignored due to non-canonical loop form from OpenMP specification
* ld: Warning: alignment 4 of symbol `KIM_CHARGE_UNIT_unused' in CMakeFiles/kim-api.dir/c/src/KIM_ChargeUnit_c.cpp.o is smaller than 8 in CMakeFiles/kim-api.dir/fortran/include/kim_charge_unit_module.f90.o (many of those, one for every atom definition)

Compiler Details from output:
`
-- <<< Compilers and Flags: >>>
-- C++ Compiler:    …/compxe.2020.1.217/compilers_and_libraries_2020.1.217/linux/bin/intel64/icpc
      Type:          Intel
      Version:       19.1.0.20200306
      C++ Flags:    -xHost -restrict 
      Defines:       LAMMPS_SMALLBIG;LAMMPS_MEMALIGN=64;LAMMPS_EXCEPTIONS;LAMMPS_OMP_COMPAT=4;LAMMPS_PNG;LAMMPS_GZIP;FFT_MKL;FFT_MKL_THREADS;LMP_PYTHON;LMP_USER_OMP;LMP_KOKKOS;LMP_USER_INTEL;LMP_INTEL_USELRT;LMP_USE_MKL_RNG
      Options:       -xHost;-O2;-no-prec-div;-qoverride-limits;-qopt-zmm-usage=high;-qno-offload;-fno-alias;-ansi-alias;-restrict
-- Fortran Compiler: …/compxe.2020.1.217/compilers_and_libraries_2020.1.217/linux/bin/intel64/ifort
      Type:          Intel
      Version:       19.1.0.20200306
      Fortran Flags: 
-- C compiler:       …/compxe.2020.1.217/compilers_and_libraries_2020.1.217/linux/bin/intel64/icc
      Type:          Intel
      Version:       19.1.0.20200306
      C Flags:      -xHost 
-- <<< Linker flags: >>>
-- Executable name:  lmp_serial
-- Static library flags:    
-- Kokkos Arch: 
-- <<< FFT settings >>>
-- Primary FFT lib:  MKL
-- Using double precision FFTs
-- Using threaded FFTs
-- Kokkos FFT: MKL
-- <<< Building Unit Tests >>>
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    ADD_TST
    PKG_FLD
    PKG_MEAM
    PKG_XTC
`

**Comment by akohlmey** [comment]
Please recall the following note from the Documentation:
```
When compiling with (aggressive) compiler optimization, some tests are likely to fail. It is recommended to inspect the individual tests in detail to decide, whether the specific error for a specific property is acceptable (it often is), or this may be an indication of mis-compiled code (or an undesired large loss of precision due to significant reordering of operations and thus less error cancellation).
```
It it quite possible that your compiler miscompiles the jacobi3 routine with high optimization (-O2 -xHost), which is crucial for correct results in this test case. So please recheck without compiler optimization (e.g. by using -DCMAKE_RUN_STYLE=Debug) or updating the compiler flags manually as needed.

We've also implemented some minor stability updates to several of the unittests (this is still a feature under active developement), so please retry with the latest patch release or master branch version from GitHub.

**Comment by iridos** [comment]
Hi, 
thanks for the extremely quick reaction.
Yes, I am of course aware that optimization can be an issue and can lead to incorrect results. 
Two points for consideration:
* lammps stable_3Mar2020  compiled without problems with the same options/compiler/mkl. 
* -O2 is chosen by cmake on my system by default. 

configuration with  -DCMAKE_RUN_STYLE=Debug:
 -- <<< Build configuration >>>
   Operating System: Linux
   Build type:       RelWithDebInfo
[…]
-- <<< Compilers and Flags: >>>
-- C++ Compiler:     /…/bin/intel64/icpc
      Type:          Intel
      Version:       19.1.0.20200306
      C++ Flags:     -restrict 
      Defines:       LAMMPS_SMALLBIG;LAMMPS_MEMALIGN=64;LAMMPS_EXCEPTIONS;LAMMPS_OMP_COMPAT=4;LAMMPS_PNG;LAMMPS_GZIP;FFT_MKL;FFT_MKL_THREADS;LMP_PYTHON;LMP_USER_OMP;LMP_KOKKOS;LMP_USER_INTEL;LMP_INTEL_USELRT;LMP_INTEL_NO_TBB;LMP_USE_MKL_RNG
      Options:       -xHost;-O2;-no-prec-div;-qoverride-limits;-qopt-zmm-usage=high;-qno-offload;-fno-alias;-ansi-alias;-restrict
-- Fortran Compiler: /…/intel64/ifort
      Type:          Intel
      Version:       19.1.0.20200306
      Fortran Flags: -O2 -g
-- C compiler:       /…/bin/intel64/icc
      Type:          Intel
      Version:       19.1.0.20200306
      C Flags:       -O2 -g -DNDEBUG
-- <<< Linker flags: >>>
-- Executable name:  lmp_serial
[…]

Result of ctest is still:
Start 6: AtomStyles
1/1 Test #6: AtomStyles .......................***Failed    0.09 sec

-------------------------

Value difference from -VV -d:
reference: 0.6250 diff: 0.1289 bonus[3].quat[0]: 0.4961 EPSILON: 5.0E-14
reference: 0.4732 diff: 0.0605 bonus[3].quat[1]: 0.5337 EPSILON: 5.0E-14
reference: 0.3307 diff: 0.1106 bonus[3].quat[2]: 0.2202 EPSILON: 5.0E-14
reference: 0.5254 diff: 0.1231 bonus[3].quat[3]: 0.6485 EPSILON: 5.0E-14

============ forcing -O0 with ccmake and changing 
=== COMPILER_SUPPORTS-O2:INTERNAL=0 in CmakeCache

-- <<< Compilers and Flags: >>>
      Version:       19.1.0.20200306
      C++ Flags:     -restrict 
      Options:       -O0;-no-prec-div;-qoverride-limits;-qopt-zmm-usage=high;-qno-offload;-fno-alias;-ansi-alias;-restrict
-- Fortran Compiler: 
      Version:       19.1.0.20200306
      Fortran Flags: -O0 -g
-- C compiler:      
      Type:          Intel
      Version:       19.1.0.20200306
      C Flags:       -O0 -g -DNDEBUG

$ ctest  -R AtomStyles
Test project /dev/shm/mqj37/chem/lammps/29Oct2020-mkl-2020-openmpi-4.0-intel-19.1/src/lammps-29Oct2020/build
    Start 6: AtomStyles
1/1 Test #6: AtomStyles .......................   Passed    0.15 sec

======== forcing -O1 
-- <<< Compilers and Flags: >>>
(same options as above, but with -O1) 

$ ctest --rerun-failed 
Test project lammps-29Oct2020/build
    Start 6: AtomStyles
1/1 Test #6: AtomStyles .......................   Passed    0.09 sec

**Comment by akohlmey** [comment]
> Hi,
> thanks for the extremely quick reaction.
> Yes, I am of course aware that optimization can be an issue and can lead to incorrect results.
> Two points for consideration:
> 
> * lammps stable_3Mar2020  compiled without problems with the same options/compiler/mkl.

but there was not unittest folder in the distribution. in fact, the body atom styles had serious issues at the time that were fixed only in the next development cycle.

> * -O2 is chosen by cmake on my system by default.

that is probably due to including the USER-INTEL package and is done in CMake code that is part of the USER-INTEL developers.

> ============ forcing -O0 with ccmake and changing
> === COMPILER_SUPPORTS-O2:INTERNAL=0 in CmakeCache
> 
> -- <<< Compilers and Flags: >>>
> Version: 19.1.0.20200306
> C++ Flags: -restrict
> Options: -O0;-no-prec-div;-qoverride-limits;-qopt-zmm-usage=high;-qno-offload;-fno-alias;-ansi-alias;-restrict
> -- Fortran Compiler:
> Version: 19.1.0.20200306
> Fortran Flags: -O0 -g
> -- C compiler:
> Type: Intel
> Version: 19.1.0.20200306
> C Flags: -O0 -g -DNDEBUG
> 
> $ ctest -R AtomStyles
> Test project /dev/shm/mqj37/chem/lammps/29Oct2020-mkl-2020-openmpi-4.0-intel-19.1/src/lammps-29Oct2020/build
> Start 6: AtomStyles
> 1/1 Test #6: AtomStyles ....................... Passed 0.15 sec
> 
> ======== forcing -O1
> -- <<< Compilers and Flags: >>>
> (same options as above, but with -O1)
> 
> $ ctest --rerun-failed
> Test project lammps-29Oct2020/build
> Start 6: AtomStyles
> 1/1 Test #6: AtomStyles ....................... Passed 0.09 sec

I am not surprised about this behavior. It is consistent with what I have experienced with the Intel compilers regularly. I makes using them a bit of a game of Russian Roulette. The good news is that this is not likely to affect the functionality tested in this specific failing test, for as long as it is consistent and then the forward and backward transformation required by the atom style will be consistent and thus not affect the final results.

In summary we can conclude that this is not an issue of the LAMMPS source code.

**Comment by iridos** [comment]
I am not compiling for myself, but for users of a HPC cluster, for which we try to offer optimized software (but of course not to the point where they produce incorrect results). What would be your suggest?

We usually run unit tests … and some example(s) to give users an impression of how to integrate with job scripts. But we do not try to cover/test all functionality beyond unit tests. That would be impossible. 

Yes, I also saw that the options are chosen by USER-INTEL.

Seems -xHost alone in conjunction with -O1 also fails the test  (below at [1], test result at[2])
Not using -xHost to accomodate for the correct architecture probably means a big performance loss, so would have to figure out what  option generated by -xHost produces the error. 

During my Phd I changes in the result due to optimization with the  intel compiler on an MD software, and decided that they probably won't change the overall result, just the specific sub-structures. But I still used a not highly optimized binary compiled with gcc for the final publication results. Maybe offering both is an option here, too. 

[1]
-- <<< Build configuration >>>
   Operating System: Linux
      C++ Flags:    -xHost -O1 -restrict 
      Options:       -xHost;-no-prec-div;-qoverride-limits;-qopt-zmm-usage=high;-qno-offload;-fno-alias;-ansi-alias;-restrict
-- Fortran Compiler: /intel64/ifort
      Type:          Intel
      Version:       19.1.0.20200306
      Fortran Flags: 
-- C compiler:       icc
      Type:          Intel
      Version:       19.1.0.20200306
      C Flags:      -xHost -O1
[2] 
ctest:
1/1 Test #6: AtomStyles .......................***Failed    0.06 sec
reference: 0.6250 diff: 0.1289 bonus[3].quat[0]: 0.4961 EPSILON: 5.0E-14
reference: 0.4732 diff: 0.0605 bonus[3].quat[1]: 0.5337 EPSILON: 5.0E-14
reference: 0.3307 diff: 0.1106 bonus[3].quat[2]: 0.2202 EPSILON: 5.0E-14
reference: 0.5254 diff: 0.1231 bonus[3].quat[3]: 0.6485 EPSILON: 5.0E-14

**Comment by akohlmey** [comment]
> I am not compiling for myself, but for users of a HPC cluster, for which we try to offer optimized software (but of course not to the point where they produce incorrect results). What would be your suggest?

It was a deliberate choice to have our unit tests be rather tight and thus only give reliable results without optimization, since their main purpose is to minimize the risk that there are no unexpected side effects from changes we integrate. LAMMPS is a large and complex software and the code quality of individual modules varies with the programming skills of the contributor and the time when it was contributed (LAMMPS is rather old and older parts of the code had to work around incomplete C++ implementations, yet we have not managed to refactor all of that). Thus you have to make a judgement call on how reliable the results are when compiled **with** optimization. If this specific test is the only failure even with compiler optimization enabled, then that is a rather good sign. As mentioned before, the failing test is about internal data and that may not even lead to inconsistent results for the user accessible data beyond the usual floating-point divergences.

If the test thresholds would be relaxed so much that also with aggressive optimization tests all tests will pass, then we would increase the risk of not detecting incorrect changes or changes with unexpected side effects.

> We usually run unit tests … and some example(s) to give users an impression of how to integrate with job scripts. But we do not try to cover/test all functionality beyond unit tests. That would be impossible.

The functionality that is used by the failing test (body particles) is a very unusual functionality, so it is highly unlikely that one of your users will come across any problems there. Mind you the LAMMPS unit tests currently cover only about 30-40% of the total code base (excluding libraries), so even with passed unit tests, there are possible issues we don't yet know about.  The aforementioned USER-INTEL package is "high risk" simply *because* of the aggressive optimizations in it. The recommendation to users would be to validate some of their results on smaller test systems without use of the USER-INTEL package code first before starting large production calculations. The same applies to about any "accelerated" package (OPT, GPU, USER-OMP, KOKKOS) in LAMMPS that provides more optimized or differently parallelized versions of the base code.

**Comment by iridos** [comment]
Ok, thank you very  much for your time, the explanation and also for the list of "accelerated packages" to exclude. I will offer two versions of lammps, one optimized with icc and one much less optimized with gcc and -O2 and without the mentioned packages,  and also   put your recommendation of how to use them into our documentation. 
Best,
Karsten
